### PR TITLE
p2p/simulations: wait until all connections are recreated when uploading snapshot

### LINKED ISSUE #1298: swarm/simulation: wait till snapshot connections are recreated
https://github.com/ethersphere/go-ethereum/pull/1284 introduces WaitTillSnapshotRecreated
this function used in simulations is blocking until all the connections specified
in the snapshot are actually up and running and show in the both node's kademlias.
It differs from WaitTillHealthy, which waits only until all the kademlias are
healthy (it might happen even before all the connections are established).

I think this functionality should be part of snapshot loading at least when using `swarm/network/simulation`, then all WaitTillHealthy calls could be eliminated from snapshot-based tests

### PR DESCRIPTION
**Pull request by gluk256**
WaitTillSnapshotRecreated is used instead of WaitTillHealthy in case snapshot was loaded from file.

closes https://github.com/ethersphere/go-ethereum/issues/1298

### CODE REVIEW

File: swarm/pss/prox_test.go
Comment (nolash) [inline swarm/pss/prox_test.go]: Nitpick; conformity - you don't have separate filename var here, like in https://github.com/ethersphere/go-ethereum/pull/1303/files#diff-fbced66e141c998cd0b7f816d30319a8R260

**Review by nolash — COMMENTED** [review]
LGTM pending tests pass

File: swarm/pss/prox_test.go
Comment (gluk256) [reply swarm/pss/prox_test.go]: fixed

File: swarm/network/stream/snapshot_retrieval_test.go
Comment (zelig) [inline swarm/network/stream/snapshot_retrieval_test.go]: No i really meant that this should be internal to `sim.UploadSnapshot` and it should block until "recreated".

**Review by zelig — CHANGES_REQUESTED** [review]
I meant that wailTillSnapshot recreated should be unexported and made part of `UploadSnapshot'

**Review by acud — CHANGES_REQUESTED** [review]
I agree with @zelig that this functionality should be internal to the package

File: swarm/network/simulation/node.go
Comment (acud) [inline swarm/network/simulation/node.go]: I'm not sure this whole error check + logging is necessary. The calling context should do the error check. Personally, I'd just `return s.WaitTillSnapshotRecreated()` in this case.

File: swarm/network/stream/snapshot_sync_test.go
Comment (acud) [inline swarm/network/stream/snapshot_sync_test.go:153]: I'd inline this in the line below

File: swarm/pss/prox_test.go
Comment (acud) [inline swarm/pss/prox_test.go:218]: this can be inlined in the line below

**Review by acud — CHANGES_REQUESTED** [review]
Minor code-style comments. Otherwise LGTM

File: swarm/network/simulation/node.go
Comment (gluk256) [reply swarm/network/simulation/node.go]: indeed. however, the logging was there in the first place, so i think i should not delete it.

File: swarm/network/stream/snapshot_sync_test.go
Comment (gluk256) [reply swarm/network/stream/snapshot_sync_test.go:153]: what is the purpose? to save a line?
now it's much more clear, IMHO.

File: swarm/pss/prox_test.go
Comment (gluk256) [reply swarm/pss/prox_test.go:218]: same as above. i suggest we discuss it with other team members. 
@frncmx : what's your opinion?

File: swarm/network/simulation/node.go
Comment (acud) [reply swarm/network/simulation/node.go]: the tests that call this function already log or return the error with a fatal exception. this can be deleted IMO but as you please

File: swarm/network/stream/snapshot_sync_test.go
Comment (acud) [reply swarm/network/stream/snapshot_sync_test.go:153]: yes you are correct, its more clear, but not uniform since in the delta below this one it is inlined. choose either one for consistency, i don't mind either one

File: swarm/network/simulation/node_test.go
Comment (frncmx) [inline swarm/network/simulation/node_test.go]: Be consistent, please. Here you write `time.Second*30`; later you write  `3*time.Minute`. The latter reads better for me; I just prefer consistency even more.

File: swarm/network/simulation/node.go
Comment (frncmx) [inline swarm/network/simulation/node.go]: Please, inline the error wherever possible. => `if err := f(); err != nil`
By keeping the variable scope minimal you don't pollute the surrounding scope and you make it clear I don't have to worry about/understand the code what comes later, I can just focus the if statement. => lower cognitive load

File: swarm/network/stream/streamer_test.go
Comment (frncmx) [inline swarm/network/stream/streamer_test.go]: For consistency if you extract `fmt.Sprintf` into a`filename` variable please make that here, too.

Going one step further this filename generation is a repeating pattern. The format string `"testing/snapshot_%d.json"` seems a little fragile and the pattern is unnecessary detail if I just want to understand the test. => Maybe worth a very simple method like `getSnapshotFilename(nodeCount int)`?

File: swarm/pss/prox_test.go
Comment (frncmx) [reply swarm/pss/prox_test.go:218]: I'm happy with the explanatory variable. As I mentioned I would maybe even make a little function instead of a variable, thus avoiding this filename generation pattern scattered and repeated everywhere.

File: swarm/network/stream/snapshot_retrieval_test.go
Comment (nolash) [inline swarm/network/stream/snapshot_retrieval_test.go:157]: 3 minutes? Why do we need that much time?

Comment (gluk256) [reply swarm/network/stream/snapshot_retrieval_test.go:157]: i thought, if there are 128 nodes and 100 messages, we might need more time.

Comment (gluk256) [reply swarm/network/stream/snapshot_retrieval_test.go:157]: you are going to change that anyway

Comment (nolash) [reply swarm/network/stream/snapshot_retrieval_test.go:157]: But it doesn't mean that it will run for 3 minutes in any case, right?

I seem to remember @nonsense being pretty vocal about tests taking too long in general in swarm, and even did quite a bit of work a while back to reduce the total amount of time tests takes. Anyway, my view is if it's a test that reasonably can take 3 minutes, then that test should be classified as `--longrunning`

Comment (gluk256) [reply swarm/network/stream/snapshot_retrieval_test.go:157]: let us refactor it after your PR is merged

Comment (nolash) [reply swarm/network/stream/snapshot_retrieval_test.go:157]: I'm not sure I understand; what will change in this particular context after my PR is merged...?

Comment (gluk256) [reply swarm/network/stream/snapshot_retrieval_test.go:157]: right. your PR overrides the timeout. no need to refactor.

Comment (nolash) [reply swarm/network/stream/snapshot_retrieval_test.go:157]: Now I understand even less :)

**Review by zelig — APPROVED** [review]

File: swarm/network/simulation/node.go
Comment (acud) [reply swarm/network/simulation/node.go]: @frncmx I don't think we should stick every assignment in an `if` statement. The accustomed syntax is to do the assignment then sequentially check for `err != nil`, I find this more readable somehow. That being said, checking for `err == nil` is something I'd like to avoid.

**Review by acud — APPROVED** [review]

File: swarm/network/simulation/node.go
Comment (frncmx) [reply swarm/network/simulation/node.go]: I agree that `if err != nil` and early return should be preferred. Good point!

File: swarm/network/simulation/node_test.go
Comment (frncmx) [inline swarm/network/simulation/node_test.go:302]: I missed `filename` extraction.

**Review by frncmx — APPROVED** [review]
lgtm

File: swarm/pss/prox_test.go
Comment (frncmx) [inline swarm/pss/prox_test.go:216]: `3*time.Minute`; just a smaller number :)

**Pull request closed without merging**
